### ISSUE TITLE
DLL Injection + MabiMods

### ISSUE THREAD

**Issue by ghost** [cursey/kanan#32]
Apparently, kanan might be taken down from MabiMods "because we don't allow injectors that can load any .dll" (quoted from ShaggyZE). This came after I uploaded a DLL injector because the one Abyss comes with has some issues.

Whole Message:

> sorry, you post was removed because we don't allow injectors that can load any .dll, which was the reason PerX and RemoteDLL was removed from Abyss and it's own Injector was needed to be made.

When inquired about Abyss and Kanan being capable of this:

> Abyss doesn't allow injecting any .dll, it has safe guards against using pake with it, and it's injector only allows it's .dll to be injected and kanan was removed once and may be removed again soon for that reason, but it hasn't been decided as of yet, but the difference is it isn't their sole purpose as patcher's to be injectors and we're not trying to promote the idea that everyone should be loading .dll's into the client, as it leads to malicious modding.

**Comment by cursey** [comment]
It's almost like calling `CreateRemoteThread` with the address of `LoadLibrary` is some big secret no one can figure out.

kanan will still be here on github for all to enjoy.  I don't like the idea of removing features legitimate users might find useful just because there _might_ be someone who uses it for wrong.

**Comment by Tonkatunk** [comment]
Mkay, as someone who was around when the maint that took abyss off-line for a while, and did what little i could to help troubleshoot things, as well as talking to some of the people involved and closely watching mabimods, i'm gonna give my viewpoint and input.

Abyss broke with a server-sided update, game never went down when abyss started randomly crashing.  I remember cuz i was finishing summon golem training and suddenly couldn't stay on 10 mins.  I also talked to some friends in discord who had clean clients and were getting kicked from the NGS detection too, once or twice.

Blade was of course informed, but guess what, all but one of his characters, an alt on a different server, were black-robed.  On top of that, his donations prior to that were pretty shitty.  He really had NO incentive to work on abyss.  But he did, and still has been, working to get abyss up and running in a user friendly manner.

During the down time, cursey/kanan came into mabimods sharing her scripts, with the focus point being bitmap fonts.  Reason being, "My friend said the game was unplayable without Bitmap font so I reversed Abyss and made this for them."  So the scripts were started with a friend in mind, on the premise of helping them have a playable game.

Between then and now, blade has gotten abyss working.  Continued work on the scripts suggests this is no longer for the individual who placed the request.

Now, abyss was reverse engineered.  This was bad.  I will give credit for at least honestly stating that though, but openly stating that and even apologizing is not the same as consent.  Being defensive on the matter, after you've essentially stolen someone's work, and playing the victim threatening to take your work off mabimods, or maybe it was just you would no longer participate in the site i don't honestly remember, is even worse.  You handled all that poorly, and i don't recall ever seeing you properly own up to it or hearing of you making amends with blade.  It leaves a bad suspicion for me, on how you've obtained other mods, whether their creators were okay with your actions.

Moving on to injecting, injecting the scripts themselves didn't seem to be much of an issue.  It's more that with all your open source work, it's quite easy for someone with minor knowledge to crack it open, get malicious ideas, and exploit the game/community after some modifications or injecting other files through your work.  Gateway threats.

Now personally, with your short but less than positive track record, what i know the mabi community would do if they had access to the more malicious content, (pake winter) and abyss being back up making most the scripts redundant, i'd once again call into question the value of leaving your scripts in mabimods at all.  

I like seeing different versions of mods out there, but with all that's tied into you and yours i see this whole thing as nothing more than a liability with an utter lack of respect for those who get victimized by your actions, and no consideration for what you're doing.  At this point in time, i only see you working for one thing, and that is yourself.

**Comment by cursey** [comment]
Hi @Tonkatunk, welcome to **my** project.  I'll cover a few points only to show you what you might have missed.

> Abyss broke with a server-sided update, game never went down when abyss started randomly crashing. I remember cuz i was finishing summon golem training and suddenly couldn't stay on 10 mins. I also talked to some friends in discord who had clean clients and were getting kicked from the NGS detection too, once or twice.
> 
> Blade was of course informed, but guess what, all but one of his characters, an alt on a different server, were black-robed. On top of that, his donations prior to that were pretty shitty. He really had NO incentive to work on abyss. But he did, and still has been, working to get abyss up and running in a user friendly manner.

This issue has nothing to do with Abyss, or Blade.  @QewQew was just informing me his contribution to the mabimods community was removed and that there were plans to do the same for my project.  I don't speak for blade's project and I do not follow it or its development.  I don't understand why you've even brought it up here.

> Now, abyss was reverse engineered. This was bad. I will give credit for at least honestly stating that though, but openly stating that and even apologizing is not the same as consent. Being defensive on the matter, after you've essentially stolen someone's work, and playing the victim threatening to take your work off mabimods, or maybe it was just you would no longer participate in the site i don't honestly remember, is even worse. You handled all that poorly, and i don't recall ever seeing you properly own up to it or hearing of you making amends with blade. It leaves a bad suspicion for me, on how you've obtained other mods, whether their creators were okay with your actions.

I'll reverse engineer anything I want. I should have gotten permission before releasing something that I made through reversing, and I did as you can see [here](https://github.com/cursey/kanan/issues/14).  I've done my best to comply with everything mabimods has asked of me. 

Also, I have not played the victim, threatened to take my work off mabimods, or said I would no longer participate on the site.  Citation needed.

> Moving on to injecting, injecting the scripts themselves didn't seem to be much of an issue. It's more that with all your open source work, it's quite easy for someone with minor knowledge to crack it open, get malicious ideas, and exploit the game/community after some modifications or injecting other files through your work. Gateway threats.

It is not my job to police the modding community into not doing something malicious.  There is no reliable way to stop it from happening anyway.

> Now personally, with your short but less than positive track record, what i know the mabi community would do if they had access to the more malicious content, (pake winter) and abyss being back up making most the scripts redundant, i'd once again call into question the value of leaving your scripts in mabimods at all.

I don't know what this malicious content is and I am sure nothing malicious is included in kanan.  If there is, its open source so please issue a PR to remove it. 

> I like seeing different versions of mods out there, but with all that's tied into you and yours i see this whole thing as nothing more than a liability with an utter lack of respect for those who get victimized by your actions, and no consideration for what you're doing. At this point in time, i only see you working for one thing, and that is yourself.

Calling **my** project redundant and then saying this makes it seem like you are _trying_ to make me and my project out to be something we're not.  I released it to help people who needed bitmap font.  Since then there have been many contributions from other people.  You can view the **public** commit log to see what I personally have authored, and what _other_ people have contributed.  I don't see how I could be working for myself when this project is 100% free and open source.  I even try documenting the features I add so that people can learn from them easily.  

In the future, when replying to an issue please make sure its relevant to the topic and less subjective.

Thanks.

**Comment by shaggyze** [comment]
Unfortunately it is our job to police the modding community, your project is redundant because abyss and mod_sharker can do the same things and more, they work fine and give credit and help us police the modding community, so they fit better into the mabimods community, so nobody needs your help with features we already have, your friend didn't need a new patcher either, he just needed patience and helping him with a quick patcher without knowing anything about mabimods was your first selfish act of working for yourself.
If your not going to be helpful to our community by limiting the chances of people using mods in a malicious manner than it only makes me think you don't give a shit of the repercussions of what can happen with an open source patcher, .dll injection and simple ideas of reverse engineering and where all those things are potentiality going to lead you and your contributors, especially if you don't get bored after you've added all the common mods, what mods do you think you'll contributors will start adding next, mods like EvGAutoRevive.js maybe? i've seen what can happen with open source patchers and i've given you the benefit of the doubt up to now, but that's a game breaking mod that mabimods will not allow it's public users to find so easily, and can only hope that you keep it private.
Seeing as QewQew aka Mr. Q is one of your contributors and decided to call me out and this discussion was ranted about in public chat instead of PM's where i had already been discussing .dll injection with Ky i see it's not just cursey that can't be trusted with what she will allow her contributors to add eventually and that eventually her contributors won't care either, but that's only my opinion and time will tell.
Blade will be fixing his Injectors issues so your failing in patience as well and to see that MabiNjector is also a project that is stepping on toes, isn't needed and also a gateway for loading malicious .dll and mods to more then just client.exe so notice how you didn't limit anything it's capable of doing, thus making it easier for people to do anything with it? if you don't understand that point then that's fine.
I'm sure epvp will love you.

**Comment by cursey** [comment]
> Unfortunately it is our job to police the modding community, your project is redundant because abyss and mod_sharker can do the same things and more, they work fine and give credit and help us police the modding community, so they fit better into the mabimods community, so nobody needs your help with features we already have, your friend didn't need a new patcher either, he just needed patience and helping him with a quick patcher without knowing anything about mabimods was your first selfish act of working for yourself.

Abyss doesn't do the same thing as kanan because it isn't open source and doesn't let you make your own mods.  mod_sharker is open source but only lets you do simple patches.  kanan lets you write scripts that can do more than just simple patches and it is open source.  I wasn't being selfish.

> If your not going to be helpful to our community by limiting the chances of people using mods in a malicious manner than it only makes me think you don't give a shit of the repercussions of what can happen with an open source patcher, .dll injection and simple ideas of reverse engineering and where all those things are potentiality going to lead you and your contributors

The reason I'm not going to limit DllLoader.js is because [there's no reason to](http://lmgtfy.com/?q=dll+injector).  Even GitHub [has plenty of projects](https://github.com/search?utf8=%E2%9C%93&q=dll+inject) that are easier to use than kanan for dll injection.  You're only kidding yourself if you think you can stop someone from figuring it out.  More importantly, there are way more legitimate users that may want to load mod_sharker or some other legitimate dll than there are malicious users and I want them to have kanan as an option.

> especially if you don't get bored after you've added all the common mods, what mods do you think you'll contributors will start adding next, mods like EvGAutoRevive.js maybe? i've seen what can happen with open source patchers and i've given you the benefit of the doubt up to now, but that's a game breaking mod that mabimods will not allow it's public users to find so easily, and can only hope that you keep it private.

Nothing in kanan makes mods like EvG auto revive any easier or harder to implement than any other project that by your own words does the same thing as kanan.  And yeah, you have to trust me that I won't release anything harmful the same way you have to trust any other mod maker on your website.  The only difference is that kanan is open source which for any other person would mean it's already more trustworthy than closed source binaries.

> Seeing as QewQew aka Mr. Q is one of your contributors and decided to call me out and this discussion was ranted about in public chat instead of PM's where i had already been discussing .dll injection with Ky i see it's not just cursey that can't be trusted with what she will allow her contributors to add eventually and that eventually her contributors won't care either, but that's only my opinion and time will tell.

The only thing that I know about @QewQew is that they contributed some open source patches to my project and fixed/improved some bits.  I'm thankful for that because that's how open source is supposed to work.  I agree that posting any private conversations publicly is in poor taste but I can't see how I'm responsible for that.  

I thought it was obvious but I'll be explicit.  **I won't accept PR's that are malicious**.

Now, what have I done specifically that is untrustworthy?

**Comment by shaggyze** [comment]
you were being selfish by releasing a patcher for a friend without regard to the patcher's currently out that were closed source to help our community maintain the possibility of malicious mods just for the sake of making your friend happy, as far as i'm aware you didn't play or have interest in mabinogi prior to that, making yours open source also gives me reason to not trust your ability to control the future of what can and can't be done with kanan and saying it has already been posted on epvp by someone other then yourself, it's going to be impossible to prevent it from being used by people that are smart enough to change it to their liking malicious or not and yes that is the way open source projects work, but with mabimods we limit what we release as open source to prevent things from getting out of hand and nexon being forced to mass ban and do roll backs as it has happened plenty of times before if we don't atleast attempt to regulate the amount of people that are helped in receiving the ability to do game breaking mods, i'm not saying it isn't possible for people to figure these things out just as easy with things other then kanan, but that doesn't mean we stop trying and just say the hell with it, do what you want and let nexon ban anyone they please that they feel are doing similar modding even if it's only for lag.
feel free to read mabimods and see how most people won't take the effort of learning .dll injection if we don't throw it into their faces and when we have to we limit it, the fact your unwilling to see or the fact it doesn't matter anymore since DllLoader.js has already been released and downloaded for those who would want or need it for modding maliciously, and yea they could be using any other injector or patcher with it, but you won't find any on mabimods which is my only point, if you wanted your mods to be posted their you would have taken the time to know what we prefer to be available to our users, since you didn't care that is what led to you not knowing how to give credit and thanks to the original developers and release mods just because your capable of reversing them, while i appreciate you learning these obvious common courtesy's of programming etiquette, your future plans for kanan are not compatible with mabimods more so now because of the ability to allow anyone to change kanan's source, inject anything and script anything without your permission or control.
it has nothing to do with not trusting you, hell i don't know you and haven't gotten the chance, but it's about not trusting people to use your work graciously and those that contribute to kanan for you or without you, simply releasing this as open source took you out of the equation which hasn't worked in your favor when having to co-op with a close community that shares mods with each other and every creator as long as they ask and talk to one another on wither it's good to release or harmful to release.

**Comment by Coziestest** [comment]
Whoa there, gang, I really wanted to stay silent and never wanted to post much, but I'm starting to feel kind of bad about Cursey having to weather this storm alone.

While we're on the subject of originality, I would like to start by first pointing out the most recent Abyss.ini file for r232:
http://pastebin.com/FAa7sw9E

Now let's go back in time and compare this to the IJL15.ini for Spr33's Fantasia patcher **circa 2009** (yes it's been that long) in the ~r50 to r60 era:
http://pastebin.com/kXuwZcnB

Do you see some similarities here?

Spr33's Fantasia patcher ultimately became the result of a collaboration with Tyne and IBWK(Kou). The IJL15.dll mods themselves were based on the IJL source code by Chris & Syoka of yydzh:
https://web.archive.org/web/20090908001243/http://disk.googleorz.com/
Let's not forget the things from jinsu either.

Spr33, before retiring from the modding scene, made Fantasia an **Open Source** IJL15.dll patcher in hopes there would be more communal effort in modding. This was a legacy that went tragically unrealized.
https://www.dropbox.com/s/buv0q4qygue5krx/fantasia_source%20(Plus%20series).zip?dl=0

Sure, for _years_ Blade has been consistent in updating it by fixing the patches whenever they break. So why didn't he let it also become open source to begin with? If he had done that he wouldn't have had to "shoulder this burden" _all by himself_ like an overworked lone wolf or something. Other members of the community might have been able to aid in fixing things when they break, or possibly even building upon it. Call me a cynic, but maybe - just maybe - he liked feeling important by being Mr. The-One-And-Only-Exclusive-Source-Of-Public-Memory-Patching-Mods. He also liked having Abyss as that easy side project that gives him a steady stream of AdFly revenue and PayPal donations out of gratitude and pity. And now he's garnering even more sympathy because it looks like he's been the victim of having a couple of "his" "original" patches stolen. But hey, I suppose at the very least he credits Spr33 for several patches in his .ini still, so he isn't a hypocrite. It's all the holier-than-thou attitudes, unwarranted self importance, thinking its their obligation to oversee the modding community, pake paranoia, and utmost effort into not being ElitePVPers in any way shape or form whatsoever which have caused people to be afraid to post new things on MabiMods. Good thing we aren't on Mabi Mods right now, eh? (All of whatever modding anyone anywhere is doing is illegal regardless)

You say Cursey is selfish? Are there profits being made here? No. Is all the glory of being the savior of the modding community in their darkest hour all being hogged by one singular individual? No. Has pake(パケ) gone unseen in the last few years since IJL15.dll became closed source? No.

Though it's been an eternity, not all of us have forgotten the legacy of Fantasia. I see Kanan as the path to the dream Spr33 envisioned. Cursey is the hero (albeit a villified one) we need, but obviously not the one we deserve. Cursey, I just hope you know that only a few are against you and the rest of us believe in and appreciate the work you do. #TeamKanan

**Comment by Tonkatunk** [comment]
For injecting dlls, and addressing the limits of abyss and mod_sharker, i'd like to point out this little work by rydian of mabimods.
https://mabimods.net/index.php?topic=16645.msg239535#msg239535

**Comment by shaggyze** [comment]
Doing something for the sake of glory of being the savior is also a selfish desire you know, and that's fine their is nothing wrong with being selfish, but you can't say it's for the sake of the modding community when their was no need for it, it's not like Abyss stopped working for a year and somebody needed to step in like in the case that Draconis started helping me update UO Tiara.
A few patchers did start off as open source, the rest are only shared with those that help code them, but most were only made open source publicly when they are dropped by their creators and new programmers stepped forward in their development making them closed source again for protection against abuse and nexon from easily patching them.
If you look at the Abyss.ini you can see he has people helping him that have the source to Abyss so practically all the mods that kanan is using are by the same mod developers that help with abyss and are capable of making the mods for people that request them, making it capable of the same things as kanan except we like to earn each other's trust and see their potential before allowing them source to our projects for them to make and add their own mods or requested ones, given that they don't cause havoc on mabinogi.
Blade isn't worried about the burden, but is more worried about his work being disregarded and used without his consent as well as things that he put in place to prevent pake being overlooked, but he does not have the sole right to be the only person making a patcher so none of us are against the idea of Kanan as long as your all aware of the potentials of it being misused.
Pake is no longer openly shared, so yea our efforts have helped.
I am only trying to beat this into your heads so that you can take some of the responsibility you've created by having it remain open source, also every other open source patcher to this day has never included dll injecting.
Obviously modding being illegal and patches being stolen being illegal or against rules isn't at question here so i don't understand the relevance, sure you can reverse engineer all you want, but you'll lose our respect and support in doing so blatantly, is it so hard to ask another programmer to be friends and take the time to earn the chance of him just telling you how it's done?
No one is complaining about being a victim or asking for sympathy, were only asking you to keep in mind the abuse that can be done which cursey already said she was going to do so that we can continue to support each other.
No one has removed kanan from mabimods since cursey added credits and made a pact with Blade  and even after you added .dll injection because yes it can be done by CE (isn't CE's main purpose), and the 1,000 other injectors that are easiliy available, but the issue is simply to give everyone fair warning about the direction kanan is headed into, you can call it paranoia, or control, but i see it as moderating potential problems even if that's impossible.

**Comment by Rydian** [comment]
Just wanted to chime in that I've been at the defense of Kanan (privately) for a while.  I've made it VERY clear to the MabiMods staff that they are being unfair.  They've (individually) brought up multiple reasons that Kanan should be removed (copying patches without prior permission, DLL loading, etc.), and every time they have I've pointed out that Abyss, my own work, or other mods also do or have done the same and should be removed as well following that logic.

MabiMods staff keeps going after Kanan because the Kanan project ticked off Blade.  Oh well.

The Kanan project is not under any obligation to deal with this.  If MabiMods decides to remove Kanan, I'll also be pulling all my work and support from MabiMods because this is NOT how a modding site should be treating contributors (and I've made it very clear that how the site treats contributors is of the utmost importance in my eyes).
